[JIT] Cast return values of functions returning Any

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42259 [JIT] Cast return values of functions returning Any**

**Summary**
This commit modifies IR generation to insert explicit cast that cast
each return value to `Any` when a function is annotated as returning `Any`.
This precludes the failure in type unification (see below) that caused
this issue.

Issue #41962 reported that the use of an `Any` return type in
combination with different code paths returning values of different
types causes a segmentation fault. This is because the exit transform
pass tries to unify the different return types, fails, but silently sets
the type of the if node to c10::nullopt. This causes problems later in
shape analysis when that type object is dereferenced.

**Test Plan**
This commit adds a unit test that checks that a function similar to the
one in #41962 can be scripted and executed.

**Fixes**
This commit fixes #41962.

Differential Revision: [D22883244](https://our.internmc.facebook.com/intern/diff/D22883244)

Commit:
[JIT] Disallow 'Any' return types for functions

**Summary**
This commit modifies the Python and C++ JIT frontends so that
they reject any function that is annotated with an `Any` return type.

Issue #41962 reported that the use of an `Any` return type in
combination with different code paths returning values of different
types causes a segmentation fault. During IR generation, the type of
each return is checked against the annotated return type of the
function. Allowing `Any` to be used as a valid return type would involve
checking that all code paths return a common supertype that is not
`Any`. If such a type can be found, it is reasonable to expect users to
use that type instead of `Any` to annotate the return type of the
function. If it cannot be found, then the code is not valid TorchScript.

Continuing to allow the use of `Any` as a return type annotation can
also mislead users into believing that code in which different code
paths return values of different types can be compiled.

**Test Plan**
This commit adds a unit test that checks that an exception with an
appropriate error message is thrown when a function and module with an
annotated return type of `Any` is compiled.

**Fixes**
This commit fixes #41962.

[ghstack-poisoned]

SplitInfinity commented on test/test_jit_py3.py:
```diff
@@ -533,6 +534,28 @@ def forward(self, x):
+            return inp * inp
+
+
+        with self.assertRaisesRegex(Exception, "Return type annotation of Any is not permitted"):
```
I tried to use `assertRaisesRegexWithHighlight` for this but it couldn't find my highlight even though I tried this out on the test case from the issue and saw that the right section was highlighted.

cc: @gmagogsfm

gmagogsfm replied:
thanks for letting me know, so just running this testcase should repro?

SplitInfinity replied:
Yes, but change this test to 
```suggestion
        with self.assertRaisesRegexWithHighlight(Exception, "Return type annotation of Any is not permitted", "Any"):
```

eellison reviewed (commented):
I don't think we should disallow this, after allowing it. It's going to break all sorts of code. I bet if you imported this there would be a bunch of failures internally. 

I think the fix is to audit all of the callsites of `unifyTypes`, there aren't that many. All of the ones outside of `ir_emitter.cpp/type.cpp` (the initial type checking) and `pybind_utils.h` should be unifying Types to `Any`. If you added another overload, `TypePtr unifyTypesWithAny` (or something along those lines) that return `Any` instead of `c10::nullopt` I think it would fix all of the failures.

Commit:
Update on "[JIT] Check mergeability of return types for functions that return Any"


**Summary**
This commit modifies the type checking performed during IR generation so
that it checks that all return types for a function marked as returning
`Any` can be unified.

Issue #41962 reported that the use of an `Any` return type in
combination with different code paths returning values of different
types causes a segmentation fault. During IR generation, the type of
each return is checked against the annotated return type of the
function, or unified with all other return types if there is no
annotation. Because every type is a subtype of `Any`, this commit
modifies the IR generation to do the latter (i.e. check if all return
types can be unified) if the annotated return type is `Any`.

**Test Plan**
This commit adds a unit test that checks that an exception with an
appropriate error message is thrown when a function and module with an
annotated return type of `Any` is compiled and the possible return types
cannot be unified.

**Fixes**
This commit fixes #41962.

[ghstack-poisoned]

SplitInfinity commented:
@eellison 

Ah, backward compatibility :'( 

It turns out that the IR emitter already checks if all return values' types can be unified if a function's return type is not explicitly annotated. I just uploaded a new version which uses this aforementioned code path when the annotated return type is `Any` instead of the code path that checks that each return value's type is a subtype of the annotated type.

Do we still need the two distinct versions of `unifyTypes` you mentioned?

eellison commented:
I think so, the updated fix still breaks BC. it's not just return types that's wrong, it's other stuff as well:
```
@torch.jit.script
def foo(cond: bool):
    x : Any = "hi"
    if cond:
        x = 3
    return x
print(foo.code)
def foo(cond: bool) -> Any:
  return "hi"
```

An output isn't getting added to the if because https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/convert_to_ssa.cpp#L95 is failing. I think we need to audit all of the use cases of unifyTypes and how they relate to possible `Any` usage, just covering up return types will still leave other errors around.

SplitInfinity commented:
I imported this into fbcode and there are no failures caused directly by this. Does that mean there are no BC issues? Or could there be some tests that will break that target determinator doesn't select?

Commit:
Update on "[JIT] Modify exit transform pass to unify types to Any"


**Summary**
This commit modifies the type unification performed during the exit
transform pass to add if node outputs so that it falls back to Any
instead of nothing (c10::nullopt). If type unification fails in this
situation, it means that the two branches of the if return different
types, so the output if the if node should be of type Any.

Issue #41962 reported that the use of an `Any` return type in
combination with different code paths returning values of different
types causes a segmentation fault. This is because the exit transform
pass tries to unify the different return types, fails, but silently sets
the type of the if node to c10::nullopt. This causes problems later in
shape analysis when that type object is dereferenced.

**Test Plan**
This commit adds a unit test that checks that a function similar to the
one in #41962 can be scripted and executed.

**Fixes**
This commit fixes #41962.

Differential Revision: [D22883244](https://our.internmc.facebook.com/intern/diff/D22883244)

[ghstack-poisoned]

SplitInfinity commented:
I think I understand what you're saying. The use of `unifyTypes` that is responsible for causing the problem in the issue is in exit_transforms.cpp:
```
  static void addIfOutputs(
      Node* n,
      at::ArrayRef<Value*> true_outs,
      at::ArrayRef<Value*> false_outs) {
    IfView if_view(n);
    registerBlockOutputs(if_view.thenBlock(), true_outs);
    registerBlockOutputs(if_view.elseBlock(), false_outs);
    for (size_t i = 0; i < true_outs.size(); ++i) {
      auto out_type =
          unifyTypes(true_outs.at(i)->type(), false_outs.at(i)->type());
      n->addOutput()->setType(*out_type);
    }
  }
```
When the two branches return values of different types, unification fails and returns `c10::nullopt`, but this code doesn't check for that and uses the type as is, leading to a segmentation fault later during shape analysis.

I updated the PR to add an optional flag to `unifyTypes` that allows the caller to force the function to fall back to `Any` instead of nothing.

Commit:
Update on "[JIT] Modify exit transform pass to unify types to Any"


**Summary**
This commit modifies the type unification performed during the exit
transform pass to add if node outputs so that it falls back to Any
instead of nothing (c10::nullopt). If type unification fails in this
situation, it means that the two branches of the if return different
types, so the output if the if node should be of type Any.

Issue #41962 reported that the use of an `Any` return type in
combination with different code paths returning values of different
types causes a segmentation fault. This is because the exit transform
pass tries to unify the different return types, fails, but silently sets
the type of the if node to c10::nullopt. This causes problems later in
shape analysis when that type object is dereferenced.

**Test Plan**
This commit adds a unit test that checks that a function similar to the
one in #41962 can be scripted and executed.

**Fixes**
This commit fixes #41962.

Differential Revision: [D22883244](https://our.internmc.facebook.com/intern/diff/D22883244)

[ghstack-poisoned]

SplitInfinity commented on test/test_jit_py3.py:
```diff
@@ -533,6 +533,22 @@ def forward(self, x):
+                return 3
+
+
+        with torch._jit_internal._disable_emit_hooks():
```
An unintended consequence of this change is that IR -> Python conversion broke because we always generate Python for an if node that looks like this:
```
if ...:
  ...
  a = 1
else:
  a = 2

return a
```

This doesn't work for this case because `a` cannot have two separate types in two separate code paths. I couldn't find a quick and easy way to modify Python code generation to account for this so I disabled export/import in the test for now while we discuss.

The only way I see around this is to hoist the return into the if and else branches if it comes immediately after the if. What do you think?

```
if ...:
  ...
  return 1
else:
  return 2
```

eellison replied:
I'm not sure what the best solution is here. Potentially we could explicitly annotate a: Any in some cases

eellison replied:
It should be possible to tell when a value is used as an Any output of an `If`, and to add an explicit annotation.

The simple case is that it only has one use, and to annotate the value with `Any` when it's printed. The trickier case is:
```
if cond:
    x = 1
    y = x + x
    return x, y
else:
    return "str", 3
```
Where the return value needs to be an int for its other uses, and then up-casted to an Any as the if-output. Maybe we could print this as 
if cond:
    x = 1
    y = x + x
    out = unchecked_cast<Any, x>, y
else:
    out = "str", 3
return out
```

SplitInfinity replied:
Is that code legal? Something like this is not:
```
def my_fn(...) -> Any:
  a: Any = 3
  if cond:
    a = 5
  else:
    a = "five"
  return a
```
but something like this is:
```
def my_func(...) -> Any:
  a: Any = 3
  if cond:
    return 5
  else:
    return "five"
```

eellison replied:
yea, you're right the first example is not legal

eellison replied:
how does that affect my comments ?

SplitInfinity replied:
I just wanted to make the point that annotating the output with `Any` is not sufficient, we need to emit the `return` inside the `if/else` branches, which we don't currently do.

eellison replied:
The problem in your example is that neither if output has Any type, which I don't htink applies to my suggestion, that's what the `unchecked_cast<Any>` is for

yf225 replied:
thanks so much for helping us on this @SplitInfinity @eellison ! Just an idea: would it help resolve the issue here if we require user to use an explicit `cast_to_any` API if they want to unify return types to `Any`? Something like:
```
def my_func(...) -> Any:
  if cond:
    return cast_to_any(5)
  else:
    return cast_to_any("five")
```

eellison replied:
@yf225 that would be a user-work around but I think we should just fix the problem in the compiler

SplitInfinity commented:
I think I managed to fix the segmentation fault from a mechanical standpoint, but I can't convince myself that this makes sense.

If the two branches of the if have corresponding outputs whose types cannot be unified without `Any`, doesn't that mean something is wrong?

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -208,7 +208,7 @@ AnyEnumTypePtr AnyEnumType::get() {
-c10::optional<TypePtr> unifyTypes(const TypePtr& t1, const TypePtr& t2) {
+c10::optional<TypePtr> unifyTypes(const TypePtr& t1, const TypePtr& t2, bool default_to_any) {
   // check direct subtyping relation
   if (t1->isSubtypeOf(t2)) {
```
All of the subcalls to unifyTypes need to thread the `default_to_any` argument

eellison reviewed (commented):
Looks good, I'm not exactly sure how we should handle the exporting... 

I think we also need to handle shape_analysis call, and the convert_to_ssa call.

SplitInfinity commented:
I have listed the callsites for `unifyTypes` below along with my thoughts on whether they should be modified or not. I'm not as familiar with the transforms so I could be wrong. Let me know what you think.

`convert_to_ssa.cpp`
```
  void addIfLoadStores(Node* n) {
  ...
    // Following the same logic as emitIfElseBlocks in ir_emitter.cpp,
    // we emit a node output if the variable is defined in each block
    // and the types of each block can be unified
    for (const auto& x : mutated_variables) {
      auto true_type = true_vars->findInAnyFrame(x);
      auto false_type = false_vars->findInAnyFrame(x);
      auto unified = unifyTypes(true_type, false_type);
      if (!unified) {
        continue;
      }

      addBlockOutput(true_block, true_type, x);
      addBlockOutput(false_block, false_type, x);
      addNodeOutput(n, *unified, x);
    }
```
Judging by the comment, it seems like this logic is supposed to mirror `ir_emitter.cpp`, which shouldn't default to `Any`.

```
      // since the loop may execute 0 or many times, the output types
      // of the loop and the input loop carried dependencies are conservatively
      // the union of the output of the body and the input to the loop
      auto block_type = loop_vars->findInThisFrame(name);
      auto unified_type = unifyTypes(parent_type, block_type).value();

      // Insert a store at the beginning of the loop block, so that all
      // loads of the variable will use the loop carried value
      addNodeInput(n, parent_type, name);
      addBlockInput(body_block, unified_type, name);
      addBlockOutput(body_block, block_type, name);
      addNodeOutput(n, unified_type, name);
```
I think we should default to `Any` here. This is basically the loop version of the change this PR makes to `exit_transforms.cpp`. We should always be able to unify the types of a loop-carried variable outside and inside a loop even if that means resorting to `Any`. However, there is no check in `ir_emitter.cpp` that the type of a loop-carried variable doesn't change inside the loop, and there is a similar check for `if` statement outputs.

`shape_analysis.cpp`
```
  bool mergeTypes(
      ArrayRef<Value*> lhs,
      ArrayRef<Value*> rhs,
      ArrayRef<Value*> outputs) {
    AT_ASSERT(lhs.size() == rhs.size() && rhs.size() == outputs.size());
    bool changed = false;
    for (size_t i = 0; i < lhs.size(); ++i) {
      auto old_output_type = outputs[i]->type();
      auto new_type = unifyTypes(lhs[i]->type(), rhs[i]->type(), /*default_to_any=*/true);
      AT_ASSERT(new_type);
      outputs[i]->setType(*new_type);
      if (*old_output_type != *outputs[i]->type())
        changed = true;
    }
    return changed;
  }
```
To be honest I have no idea what this code is for but given that it expects `new_type` to always be valid, it seems like defaulting to `Any` is the right thing to do?

Commit:
Update on "[JIT] Cast return values of functions returning Any"


**Summary**
This commit modifies IR generation to insert explicit cast that cast
each return value to `Any` when a function is annotated as returning `Any`.
This precludes the failure in type unification (see below) that caused
this issue.

Issue #41962 reported that the use of an `Any` return type in
combination with different code paths returning values of different
types causes a segmentation fault. This is because the exit transform
pass tries to unify the different return types, fails, but silently sets
the type of the if node to c10::nullopt. This causes problems later in
shape analysis when that type object is dereferenced.

**Test Plan**
This commit adds a unit test that checks that a function similar to the
one in #41962 can be scripted and executed.

**Fixes**
This commit fixes #41962.

Differential Revision: [D22883244](https://our.internmc.facebook.com/intern/diff/D22883244)

[ghstack-poisoned]

SplitInfinity commented:
@eellison @yf225 

I made the changes we discussed and this particular issue seems to be solved. I opened #43378 for the other problem we discussed because I think it is sufficiently distinct from this one.

Commit:
Update on "[JIT] Cast return values of functions returning Any"


**Summary**
This commit modifies IR generation to insert explicit cast that cast
each return value to `Any` when a function is annotated as returning `Any`.
This precludes the failure in type unification (see below) that caused
this issue.

Issue #41962 reported that the use of an `Any` return type in
combination with different code paths returning values of different
types causes a segmentation fault. This is because the exit transform
pass tries to unify the different return types, fails, but silently sets
the type of the if node to c10::nullopt. This causes problems later in
shape analysis when that type object is dereferenced.

**Test Plan**
This commit adds a unit test that checks that a function similar to the
one in #41962 can be scripted and executed.

**Fixes**
This commit fixes #41962.

Differential Revision: [D22883244](https://our.internmc.facebook.com/intern/diff/D22883244)

[ghstack-poisoned]

yf225 commented:
Thanks so much @SplitInfinity! Curious do you mind checking if the following use case works as well?
```python
import torch
from typing import Any, Dict

class IdListFeature(object):
    def __init__(self, lengths, values):
        self.lengths = lengths
        self.values = values


class IdScore(object):
    def __init__(self, ids, scores):
        self.ids = ids
        self.scores = scores


class IdScoreListFeature(object):
    def __init__(self, lengths, ids, scores):
        self.lengths = lengths
        self.values = IdScore(ids=ids, scores=scores)


class HashFeatureIds(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, module_input: Any) -> Any:
        if isinstance(module_input, IdListFeature):
            return module_input

        if isinstance(module_input, IdScoreListFeature):
            return module_input

        raise Exception

m = HashFeatureIds()
torch.jit.script(m)
```
The only difference is that the return value is now actually a custom TorchScript class. Thanks again!

SplitInfinity commented:
@yf225 

It scripts! 🥳

eellison reviewed (commented):
Looks good, just needs one fix re: shape_analysis. 

As https://github.com/pytorch/pytorch/issues/43378 shows, we still have pretty significant holes in our `Any` support that need fixing. 

One thing I don't like about this approach is that it's very fragile to optimizations - if we run peephole optimizations on the graph then that will break export. I suspect with the fix to https://github.com/pytorch/pytorch/issues/43378 we will have to revisit our approach here and when we unify types to Any. 

Nevertheless i think this does make code base better and fixes a big use case so i think we should land this, we just need to update shape_analysis.

Commit:
Update on "[JIT] Cast return values of functions returning Any"


**Summary**
This commit modifies IR generation to insert explicit cast that cast
each return value to `Any` when a function is annotated as returning `Any`.
This precludes the failure in type unification (see below) that caused
this issue.

Issue #41962 reported that the use of an `Any` return type in
combination with different code paths returning values of different
types causes a segmentation fault. This is because the exit transform
pass tries to unify the different return types, fails, but silently sets
the type of the if node to c10::nullopt. This causes problems later in
shape analysis when that type object is dereferenced.

**Test Plan**
This commit adds a unit test that checks that a function similar to the
one in #41962 can be scripted and executed.

**Fixes**
This commit fixes #41962.

Differential Revision: [D22883244](https://our.internmc.facebook.com/intern/diff/D22883244)

[ghstack-poisoned]

Commit:
Update on "[JIT] Cast return values of functions returning Any"


**Summary**
This commit modifies IR generation to insert explicit cast that cast
each return value to `Any` when a function is annotated as returning `Any`.
This precludes the failure in type unification (see below) that caused
this issue.

Issue #41962 reported that the use of an `Any` return type in
combination with different code paths returning values of different
types causes a segmentation fault. This is because the exit transform
pass tries to unify the different return types, fails, but silently sets
the type of the if node to c10::nullopt. This causes problems later in
shape analysis when that type object is dereferenced.

**Test Plan**
This commit adds a unit test that checks that a function similar to the
one in #41962 can be scripted and executed.

**Fixes**
This commit fixes #41962.

Differential Revision: [D22883244](https://our.internmc.facebook.com/intern/diff/D22883244)

[ghstack-poisoned]

eellison commented on aten/src/ATen/core/type.cpp:
```diff
@@ -287,6 +287,16 @@ c10::optional<TypePtr> unifyTypes(const TypePtr& t1, const TypePtr& t2) {
+  auto unified = unifyTypesImpl(t1, t2);
+
+  if (default_to_any && !unified) {
+    return c10::optional<TypePtr>(AnyType::get());
```
nit: i dont think you actually need the cast here, you only needed the cast for pybind bc it was auto-inferring the return type

eellison commented on paste:
```diff
@@ -0,0 +1,38 @@
+@silent do docker ps and then either docker attach <container id> or, if that does not work, docker run -ti <image_name> /bin/bash
```
Accidental include

eellison commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1638,7 +1638,8 @@ inline at::ScalarType scalarTypeFromJitType(const c10::TypePtr& type) {
     const TypePtr& t1,
-    const TypePtr& t2);
+    const TypePtr& t2,
+    bool default_to_any = false);
```
default to any is somewhat self-explanatory but maybe worth revising the comments above

eellison reviewed (approved):
LGTM! you have an accidental file included

Commit:
Update on "[JIT] Cast return values of functions returning Any"


**Summary**
This commit modifies IR generation to insert explicit cast that cast
each return value to `Any` when a function is annotated as returning `Any`.
This precludes the failure in type unification (see below) that caused
this issue.

Issue #41962 reported that the use of an `Any` return type in
combination with different code paths returning values of different
types causes a segmentation fault. This is because the exit transform
pass tries to unify the different return types, fails, but silently sets
the type of the if node to c10::nullopt. This causes problems later in
shape analysis when that type object is dereferenced.

**Test Plan**
This commit adds a unit test that checks that a function similar to the
one in #41962 can be scripted and executed.

**Fixes**
This commit fixes #41962.

Differential Revision: [D22883244](https://our.internmc.facebook.com/intern/diff/D22883244)

[ghstack-poisoned]
